src/lib/Persistence/Legacy/Content/Mapper/ResolveVirtualFieldSubscriber.php

```diff
@@ -0,0 +1,215 @@
+                $content->versionInfo,
+                $event->getFieldDefinition(),
+                $event->getLanguageCode()
+            );
+
+            $event->setField($emptyField);
+        } catch (NotFound $exception) {
+            return;
+        }
+    }
+
+    /**
+     * @throws \eZ\Publish\Core\Persistence\Legacy\Content\FieldValue\Converter\Exception\NotFound
+     */
+    public function persistExternalStorageField(ResolveMissingFieldEvent $event): void
+    {
+        $field = $event->getField();
+
+        if ($field && $field->id !== null) {
+            // Not a virtual field
+            return;
+        }
+
+        $fieldDefinition = $event->getFieldDefinition();
+        $storage = $this->storageRegistry->getStorage($fieldDefinition->fieldType);
+
+        if ($storage instanceof NullStorage) {
+            // Not an external storage
+            return;
+        }
+
+        $content = $event->getContent();
+
+        $field->id = $this->contentGateway->insertNewField(
+            $content,
+            $field,
+            $this->getDefaultStorageValue()
+        );
+
+        if ($field->value->data !== null) {
+            $result = $storage->storeFieldData(
+                $content->versionInfo,
+                $field,
+                []
+            );
+
+            if ($result === true) {
+                $storageValue = new StorageFieldValue();
+                $converter = $this->converterRegistry->getConverter($fieldDefinition->fieldType);
+                $converter->toStorageValue(
+                    $field->value,
+                    $storageValue
+                );
+
+                $this->contentGateway->updateField(
+                    $field,
+                    $storageValue
+                );
+            }
+        }
+
+        $storage->getFieldData(
+            $content->versionInfo,
+            $field,
+            []
+        );
+
+        $event->setField($field);
+    }
+
+    public function resolveVirtualExternalStorageField(ResolveMissingFieldEvent $event): void
+    {
+        $field = $event->getField();
+
+        if ($field && $field->id !== null) {
+            // Not a virtual field
+            return;
+        }
+
+        $fieldDefinition = $event->getFieldDefinition();
+        $storage = $this->storageRegistry->getStorage($fieldDefinition->fieldType);
+
+        if ($storage instanceof NullStorage) {
+            // Not an external storage
+            return;
+        }
+
+        if (!$storage instanceof DefaultDataFieldStorage) {
+            return;
+        }
+
+        $content = $event->getContent();
+
+        $storage->getDefaultFieldData(
+            $content->versionInfo,
+            $field
+        );
+
+        $event->setField($field);
+
+        // Do not persist the external storage field
+        $event->stopPropagation();
+    }
+
+    /**
+     * @throws \eZ\Publish\Core\Persistence\Legacy\Content\FieldValue\Converter\Exception\NotFound
```

Comment:
We rather avoid reporting non-contract exceptions:
```suggestion
     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
```